share/wake/lib/core/lens.wake

```diff
@@ -0,0 +1,156 @@
+# Copyright 2021 SiFive, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You should have received a copy of LICENSE.Apache2 along with
+# this software. If not, you may obtain a copy at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+package wake
+
+export def (l: List a) |> (t: a => List b): List b = mapFlat t l
+
+# TODO: How do we feel about `x | f |> g` vs `|> x |> f |> g`?
+export def |> (x: a): List a = x, Nil
```

Comment:
I'm ambivalent.

Reply:
I think it's a bit weird to use the same name for both the "return" and "bind" functions of a Monad. I personally think it's better to use different names, but that's just my opinion.

Reply:
So I've been hitting the need for the ability to make singelton lists a lot and I don't like using `, Nil`, for the purposes of this change I can just remove this operator but I think I'd like *some* way to do this. I don't want to call it "pure" since its List specific and I think "pure/return" will confuse people a lot. "singelton" is what Haskell calls this in Data.List but that's *really* long to type out in the places I need it. '|>' interacts optimally with the precedence rules we need but so would anything else with '|' at the start.

Any other suggestions?

Reply:
I do agree that it looks weird when written on a single line like this, but when everything's properly linebroken, it makes more visual sense.  My only concern is that having a unary and a binary operator which are written identically but do different (if related) things might wind up a bit confusing, so I'll cast a tentative vote for some alphanumeric name.  Agreed on not liking `pure` given the specificity, and in general feel that Wake is a bit wordier than Haskell so I wouldn't mind `singleton`, but since you're the one who's actually been writing it, what about abbreviating to `single`?

Reply:
So the proposal is?:
```
require x, Nil =
  single json
  |> jField "abc"
  |> jField "xyz"
```
... I prefer this to `json | jField |> jField`. I think I even prefer it to `|>`. So, I am no longer ambivalent. I vote for @ag-eitilt 's proposal.

Reply:
I Sam's proposal too, I'll add this.

Reply:
That is indeed the proposal, and thanks for putting the example together!  Sounds like we're agreed then.

Reply:
BTW, if we ever add syntactic patterns, this becomes legal:

```
require single x =
   ...
```